### PR TITLE
add `accelerator.is_available()` check

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

part of #11449:  Enable accelerator.is_available() check
reference: implemented in PR #11797

### Does your PR introduce any breaking changes? If yes, please list them.

adds hardware check in Trainer (AcceleratorConnector) init

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 1]
```diff
@@ -492,7 +492,8 @@ def _set_parallel_devices_and_init_accelerator(self) -> None:
                     "The flag `devices` must be an int with `accelerator='cpu'`,"
                     f" got `devices={self._devices_flag}` instead."
                 )
-
+        if not self.accelerator.is_available():
+            raise MisconfigurationException(f"{self.accelerator.__class__.__qualname__} can not run on this hardware.")
```
Comment (four4fish) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:496]: Could we change the exception to X devices is not available? I think users are more familiar with that

**Review by four4fish — APPROVED** [review]
LGTM! Thank you for adding this

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 2]
```diff
@@ -493,7 +493,13 @@ def _set_parallel_devices_and_init_accelerator(self) -> None:
...
                 )
         if not self.accelerator.is_available():
-            raise MisconfigurationException(f"{self.accelerator.__class__.__qualname__} can not run on this hardware.")
+            if isinstance(self.accelerator, GPUAccelerator):
+                hardware_str = "GPU"
+            elif isinstance(self.accelerator, IPUAccelerator):
+                hardware_str = "IPU"
+            else:
+                hardware_str = "TPU"
+            raise MisconfigurationException(f"{self.accelerator.__class__.__qualname__} can not run on this hardware, " + hardware_str + "s are not available.")
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py]: - the error message should show what hardware is available
- the error message should show how uses can configure this on the Trainer (e.g. `Trainer(accelerator="...")`

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 3]
```diff
@@ -492,7 +492,18 @@ def _set_parallel_devices_and_init_accelerator(self) -> None:
                     "The flag `devices` must be an int with `accelerator='cpu'`,"
                     f" got `devices={self._devices_flag}` instead."
                 )
-
+        if not self.accelerator.is_available():
+            if isinstance(self.accelerator, GPUAccelerator):
+                hardware_str = "GPU"
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:497]: Won't be better to add hardware_str to the Accelerator directly?

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 4]
```diff
@@ -492,7 +492,18 @@ def _set_parallel_devices_and_init_accelerator(self) -> None:
...
+            elif isinstance(self.accelerator, IPUAccelerator):
+                hardware_str = "IPU"
+            else:
+                hardware_str = "TPU"
+            raise MisconfigurationException(
+                f"{self.accelerator.__class__.__qualname__} can not run on this hardware, "
+                + hardware_str
+                + "s are not available."
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:505]:
```suggestion
                f"{self.accelerator.__class__.__qualname__} can not run on this hardware, {hardware_str}s are not"
                " available."
```

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:497]: +1, I would expect to have a `hardware_name` just as we do for `strategy_name`.

Fine doing this in a follow-up if you prefer

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 5]
```diff
@@ -492,7 +492,18 @@ def _set_parallel_devices_and_init_accelerator(self) -> None:
...
                     f" got `devices={self._devices_flag}` instead."
                 )
-
+        if not self.accelerator.is_available():
+            if isinstance(self.accelerator, GPUAccelerator):
+                hardware_str = "GPU"
+            elif isinstance(self.accelerator, IPUAccelerator):
+                hardware_str = "IPU"
+            else:
+                hardware_str = "TPU"
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:501]: To match `strategy_name`

```suggestion
                hardware_name = "TPU"
```

**Review by carmocca — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]
LGTM. Once others' suggestions are addressed.

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 2 above)
Comment (jjenniferdai) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py]: updated msg

File: pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 3 above)
Comment (jjenniferdai) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py]: @tchaton @carmocca updated, is it similar to what you had in mind?

File: pytorch_lightning/accelerators/cpu.py [hunk 6]
```diff
@@ -62,3 +62,8 @@ def auto_device_count() -> int:
     def is_available() -> bool:
         """CPU is always available for execution."""
         return True
+
+    @staticmethod
+    def hardware_name() -> str:
```
Comment (kaushikb11) [inline pytorch_lightning/accelerators/cpu.py:67]: I would prefer the name `accelerator_type`

File: tests/trainer/properties/test_estimated_stepping_batches.py [hunk 7]
```diff
@@ -148,7 +149,8 @@ def test_num_stepping_batches_with_tpu(devices, estimated_steps):
     assert trainer.estimated_stepping_batches == estimated_steps
 
 
-def test_num_stepping_batches_with_ipu(monkeypatch):
+@mock.patch("pytorch_lightning.accelerators.ipu.IPUAccelerator.is_available", return_value=True)
+def test_num_stepping_batches_with_ipu(mock_ipuacc_avail, monkeypatch):
```
Comment (ananthsub) [inline tests/trainer/properties/test_estimated_stepping_batches.py:153]:
```suggestion
def test_num_stepping_batches_with_ipu(mock_ipu_acc_avail, monkeypatch):
```

File: tests/plugins/test_cluster_integration.py [hunk 8]
```diff
@@ -58,7 +58,8 @@ def environment_combinations():
     "strategy_cls",
     [DDPStrategy, DDPShardedStrategy, DDP2Strategy, pytest.param(DeepSpeedStrategy, marks=RunIf(deepspeed=True))],
 )
-def test_ranks_available_manual_strategy_selection(strategy_cls):
+@mock.patch("pytorch_lightning.accelerators.gpu.GPUAccelerator.is_available", return_value=True)
+def test_ranks_available_manual_strategy_selection(mock_gpuacc_available, strategy_cls):
```
Comment (ananthsub) [inline tests/plugins/test_cluster_integration.py:62]:
```suggestion
def test_ranks_available_manual_strategy_selection(mock_gpu_acc_available, strategy_cls):
```

File: tests/accelerators/test_ipu.py [hunk 9]
```diff
@@ -97,7 +98,8 @@ def test_epoch_end(self, outputs) -> None:
 
 
 @pytest.mark.skipif(_IPU_AVAILABLE, reason="test requires non-IPU machine")
-def test_fail_if_no_ipus(tmpdir):
+@mock.patch("pytorch_lightning.accelerators.ipu.IPUAccelerator.is_available", return_value=True)
+def test_fail_if_no_ipus(mock_ipuacc_avail, tmpdir):
```
Comment (ananthsub) [inline tests/accelerators/test_ipu.py:102]:
```suggestion
def test_fail_if_no_ipus(mock_ipu_acc_avail, tmpdir):
```

File: tests/accelerators/test_accelerator_connector.py [hunk 10]
```diff
@@ -914,7 +917,8 @@ def test_unsupported_tpu_choice(mock_devices):
             Trainer(accelerator="tpu", precision=16, amp_backend="apex", strategy="single_device")
 
 
-def test_unsupported_ipu_choice(monkeypatch):
+@mock.patch("pytorch_lightning.accelerators.ipu.IPUAccelerator.is_available", return_value=True)
+def test_unsupported_ipu_choice(mock_ipuacc_avail, monkeypatch):
```
Comment (ananthsub) [inline tests/accelerators/test_accelerator_connector.py:921]:
```suggestion
def test_unsupported_ipu_choice(mock_ipu_acc_avail, monkeypatch):
```

File: tests/accelerators/test_accelerator_connector.py [hunk 11]
```diff
@@ -898,8 +900,9 @@ def test_strategy_choice_ddp_cpu_slurm(device_count_mock, setup_distributed_mock
     assert trainer.strategy.local_rank == 0
 
 
+@mock.patch("pytorch_lightning.accelerators.tpu.TPUAccelerator.is_available", return_value=True)
 @mock.patch("pytorch_lightning.accelerators.tpu.TPUAccelerator.parse_devices", return_value=8)
-def test_unsupported_tpu_choice(mock_devices):
+def test_unsupported_tpu_choice(mock_devices, mock_tpuacc_avail):
```
Comment (ananthsub) [inline tests/accelerators/test_accelerator_connector.py:905]:
```suggestion
def test_unsupported_tpu_choice(mock_devices, mock_tpu_acc_avail):
```

File: tests/accelerators/test_accelerator_connector.py [hunk 12]
```diff
@@ -428,8 +428,9 @@ def test_ipython_compatible_dp_strategy_gpu(_, monkeypatch):
     assert trainer.strategy.launcher is None or trainer.strategy.launcher.is_interactive_compatible
 
 
+@mock.patch("pytorch_lightning.accelerators.tpu.TPUAccelerator.is_available", return_value=True)
 @mock.patch("pytorch_lightning.accelerators.tpu.TPUAccelerator.parse_devices", return_value=8)
-def test_ipython_compatible_strategy_tpu(_, monkeypatch):
+def test_ipython_compatible_strategy_tpu(mock_devices, mock_tpuacc_avail, monkeypatch):
```
Comment (ananthsub) [inline tests/accelerators/test_accelerator_connector.py:433]:
```suggestion
def test_ipython_compatible_strategy_tpu(mock_devices, mock_tpu_acc_avail, monkeypatch):
```

File: pytorch_lightning/accelerators/cpu.py (hunk 6 above)
Comment (kaushikb11) [reply pytorch_lightning/accelerators/cpu.py:67]: > if we want this to work with auto, then it should be a abstractmethod staticmethod on the base interface.

Absolutely! It should be part of base class as abstractmethod as well.

> i prefer name here to reduce verbosity

We have `strategy_name` attribute as part of Strategy. I think we could do `accelerator_name` as well.

File: pytorch_lightning/accelerators/cpu.py [hunk 13]
```diff
@@ -62,3 +62,8 @@ def auto_device_count() -> int:
...
+        """Name of the hardware."""
+        return "CPU"
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply pytorch_lightning/accelerators/cpu.py:69]: +1

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 14]
```diff
@@ -455,16 +455,15 @@ def _choose_accelerator(self) -> str:
         return "cpu"
 
     def _set_parallel_devices_and_init_accelerator(self) -> None:
-        # TODO add device availability check
+        ACCELERATORS = {
+            "cpu": CPUAccelerator,
+            "gpu": GPUAccelerator,
+            "tpu": TPUAccelerator,
+            "ipu": IPUAccelerator,
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:462]: Yuss! We also need AcceleratorRegistry for #12030. I will add it once this PR is merged.

File: pytorch_lightning/accelerators/cpu.py (hunk 6 above)
Comment (ananthsub) [reply pytorch_lightning/accelerators/cpu.py]: can we rename `strategy_name` to be simply `name` ? `strategy_name` was only recently introduced in the accelerator connector rewrite, so it's safe to change immediately w/o concerns of breaking backward compatibility. @four4fish

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 15]
```diff
@@ -455,16 +455,15 @@ def _choose_accelerator(self) -> str:
...
+        }
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:463]: currently this duplicates the name implementation of each accelerator class here. this is a risk in case the name ever changes (unlikely to happen, but still). this would've protected against that:
```py
acc_cls = (CPUAccelerator, GPUAccelerator, TPUAccelerator, IPUAccelerator)
ACCELERATORS = {acc_cls.name(): acc_cls for acc in acc_cls}
```

Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/accelerator_connector.py:463]: I think this will be removed here: https://github.com/PyTorchLightning/pytorch-lightning/pull/12180

File: pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 16]
```diff
@@ -475,6 +474,15 @@ def _set_parallel_devices_and_init_accelerator(self) -> None:
             accelerator_class = ACCELERATORS[self._accelerator_flag]
             self.accelerator = accelerator_class()  # type: ignore[abstract]
 
+        if not self.accelerator.is_available():
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/accelerator_connector.py:477]: depending on the accelerator, we might not be able to instantiate it without it being available. The init should have the freedom to assume it is available or raise additional errors. IMO the check should have been before the instantiation.
